messaging/src/main/java/org/axonframework/eventhandling/MultiSourceTrackingToken.java

```diff
@@ -0,0 +1,194 @@
+     * token
+     *
+     * @param other The token to compare to this one
+     * @return token representing the lower bound of of both tokens
+     */
+    @Override
+    public TrackingToken lowerBound(TrackingToken other) {
+        Assert.isTrue(other instanceof MultiSourceTrackingToken, () -> "Incompatible token type provided.");
+
+        MultiSourceTrackingToken otherMultiToken = (MultiSourceTrackingToken) other;
+
+        Assert.isTrue(otherMultiToken.trackingTokens.keySet().equals(this.trackingTokens.keySet()),
+                      () -> "MultiSourceTrackingTokens contain different keys");
+
+        Map<String, TrackingToken> tokenMap = new HashMap<>();
+
+        otherMultiToken.trackingTokens.forEach((key, otherToken) ->
+                                                       tokenMap.put(key, trackingTokens.get(key).lowerBound(otherToken))
+        );
+
+        return new MultiSourceTrackingToken(tokenMap);
+    }
+
+    /**
+     * Compares this token to {@code other} by comparing each member token with its counterpart in the {@code other}
+     * token
+     *
+     * @param other The token to compare this token to
+     * @return a token that represents the furthest position of this or the other streams
+     */
+    @Override
+    public TrackingToken upperBound(TrackingToken other) {
+        Assert.isTrue(other instanceof MultiSourceTrackingToken, () -> "Incompatible token type provided.");
+
+        MultiSourceTrackingToken otherMultiToken = (MultiSourceTrackingToken) other;
+
+        Assert.isTrue(otherMultiToken.trackingTokens.keySet().equals(this.trackingTokens.keySet()),
+                      () -> "MultiSourceTrackingTokens contain different keys");
+
+        Map<String, TrackingToken> tokenMap = new HashMap<>();
+
+        otherMultiToken.trackingTokens.forEach((key, otherToken) ->
+                                                       tokenMap.put(key, trackingTokens.get(key).upperBound(otherToken))
+        );
+
+        return new MultiSourceTrackingToken(tokenMap);
+    }
+
+    /**
+     * Compares this token to {@code other} checking each member token with its counterpart to see if they are covered
+     * in the {@code other} token
+     *
+     * @param other The token to compare to this one
+     * @return {@code true} if this token covers the other, otherwise {@code false}
+     */
+    @Override
+    public boolean covers(TrackingToken other) {
+        Assert.isTrue(other instanceof MultiSourceTrackingToken, () -> "Incompatible token type provided.");
+
+        MultiSourceTrackingToken otherMultiToken = (MultiSourceTrackingToken) other;
+
+        Assert.isTrue(otherMultiToken.trackingTokens.keySet().equals(this.trackingTokens.keySet()),
+                      () -> "MultiSourceTrackingTokens contain different keys");
+
+        //as soon as one delegated token doesn't cover return false
+        for (Map.Entry<String, TrackingToken> trackingTokenEntry : trackingTokens.entrySet()) {
```

Comment:
Suggestion: You could rewrite this to:
```
return trackingTokens.entrySet().stream().noneMatch(
        trackingTokenEntry -> trackingTokenEntry.getValue().covers(
                otherMultiToken.trackingTokens.get(trackingTokenEntry.getKey())
        )
);
```

Doing it like this, will also remove the pesky small negation by the exclamation mark, which in a lot of scenarios is missed.